bip-0360.mediawiki

```diff
@@ -167,389 +171,305 @@ HASH160 <ref name="hash160">Used by P2PKH, P2SH, and P2WPKH addresses, though no
+to compute the tapscript merkle root. However, instead of the root of the Merkle tree being hashed together with the internal
+key in P2QRH the root is hashed by itself using the tag "QuantumRoot" and then set as the witness program.
 
-=== Script Validation ===
+=== Address Format ===
 
-To spend a P2QRH output, the following conditions must be met:
+P2QRH uses SegWit version 3 outputs, resulting in addresses that start with <code>bc1r</code>, following
+[https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32 BIP-173]. Bech32 encoding maps version 3 to the
+prefix <code>r</code>.
 
-1. The <code>scriptPubKey</code> must be of the form:
+Example P2QRH address:
 
-OP_PUSHNUM_3 <32-byte hash>
+<code>bc1r...</code> (32-byte Bech32m-encoded tapleaf merkle root)
 
-2. The attestation must include:
+=== ScriptPubKey ===
 
-* The quantum-resistant public key(s) whose HASH256 concatenated and hashed again matches the <nowiki><hash></nowiki> in
-the <code>scriptPubKey</code>.
+The <code>scriptPubKey</code> for a P2QRH output is:
 
-* Valid signatures corresponding to the public key(s) and the transaction data.
+  OP_PUSHNUM_3 OP_PUSHBYTES_32 <nowiki><hash></nowiki>
 
-* The key type bitmask and threshold must match the commitment in the <code>scriptPubKey</code>.
+Where:
 
-3. For multi-signature schemes, all required public keys and signatures must be provided for that input within the
-attestation. Public keys that are not needed or available can be selectively disclosed by including their hash in the
-attestation accompanied with an empty signature by providing a 0x00 signature length byte. This works so long as
-enough keys to meet the threshold are provided.
+* <code>OP_PUSHNUM_3</code> (<code>0x53</code>) indicates SegWit version 3.
+* <nowiki><hash></nowiki> is the 32-byte tapleaf merkle root.
+
+==== Script Validation ====
+
+A P2QRH output is a native SegWit output (see [[bip-0141.mediawiki|BIP141]]) with version number 3, and a 32-byte witness program.
+Unlike taproot this witness program is the tapleaf merkle root. For the sake of comparison we have, as much as possible, copied the 
+language verbatim from the [[bip-0341.mediawiki|BIP341]] script validation section.
+
+* Let ''q'' be the 32-byte array containing the witness program (the second push in the scriptPubKey) which represents root of tapleaf merkle tree.
+* Fail if the witness stack does not have two or more elements.
+* If there are at least three witness elements, and the first byte of the last element is 0x50, this last element is called ''annex'' ''a'' and is removed from the witness stack. The annex (or the lack of thereof) is always covered by the signature and contributes to transaction weight, but is otherwise ignored during taproot validation.
+* There must be at least two witness elements left.
+** Call the last stack element ''s'', the script (as defined in [[bip-0341.mediawiki|BIP341]])
```

Comment:
second-to-last stack element, right? The next bullet also calls the last stack element the control block.

bip-341 for reference has the second-to-last element as the script and the last element the control block. (This ordering also makes sense so that the first byte of the control block can be distinguished from the annex marker byte)

Reply:
That's my mistake. Not sure how that got changed. Fixed